src/marks/waffle.js

```diff
@@ -0,0 +1,202 @@
+import {extent, namespaces} from "d3";
+import {create} from "../context.js";
+import {composeRender} from "../mark.js";
+import {hasXY, identity, indexOf} from "../options.js";
+import {applyChannelStyles, applyDirectStyles, applyIndirectStyles, getPatternId} from "../style.js";
+import {template} from "../template.js";
+import {maybeIdentityX, maybeIdentityY} from "../transforms/identity.js";
+import {maybeIntervalX, maybeIntervalY} from "../transforms/interval.js";
+import {maybeStackX, maybeStackY} from "../transforms/stack.js";
+import {BarX, BarY} from "./bar.js";
+
+const waffleDefaults = {
+  ariaLabel: "waffle"
+};
+
+export class WaffleX extends BarX {
+  constructor(data, {unit = 1, gap = 1, round, render, ...options} = {}) {
+    super(data, {...options, render: composeRender(render, waffleRender("x"))}, waffleDefaults);
+    this.unit = Math.max(0, unit);
+    this.gap = +gap;
+    this.round = maybeRound(round);
+  }
+}
+
+export class WaffleY extends BarY {
+  constructor(data, {unit = 1, gap = 1, round, render, ...options} = {}) {
+    super(data, {...options, render: composeRender(render, waffleRender("y"))}, waffleDefaults);
+    this.unit = Math.max(0, unit);
+    this.gap = +gap;
+    this.round = maybeRound(round);
+  }
+}
+
+function waffleRender(y) {
+  return function (index, scales, values, dimensions, context) {
+    const {unit, gap, rx, ry, round} = this;
+    const {document} = context;
+    const Y1 = values.channels[`${y}1`].value;
+    const Y2 = values.channels[`${y}2`].value;
+
+    // We might not use all the available bandwidth if the cells don’t fit evenly.
+    const barwidth = this[y === "y" ? "_width" : "_height"](scales, values, dimensions);
+    const barx = this[y === "y" ? "_x" : "_y"](scales, values, dimensions);
+
+    // The length of a unit along y in pixels.
+    const scale = unit * scaleof(scales.scales[y]);
+
+    // The number of cells on each row of the waffle.
+    const columns = Math.max(1, Math.floor(Math.sqrt(barwidth / scale)));
+
+    // The outer size of each square cell, in pixels, including the gap.
+    const cellsize = scale * columns;
```

Comment:
```suggestion
    const cellsize = Math.min(barwidth, scale * columns);
```

Clamping `cellsize` to `barwidth` fixes the situation where waffles overlap visually, typically when you only have low values: `Plot.waffleY([1, 1, 1, 1]).plot()`. It cheats a bit with the *y* axis but I think it makes a more readable chart.

| before | after |
| - | - |
| ![before](https://github.com/user-attachments/assets/06764888-4728-4087-a5c3-de99a035efca) | ![after](https://github.com/user-attachments/assets/2eaaf901-2225-4857-a0e5-c9f683f7848e) |

Reply:
Unfortunately, you can’t satisfy both constraints: you can’t have both square waffle cells and avoid overlap. I think it’s more than “it cheats a bit” — it’s grossly misleading. For example here the value 10 looks like ~4:

<img width="644" alt="Screenshot 2024-08-04 at 3 01 09 PM" src="https://github.com/user-attachments/assets/417ac826-d776-41c8-98c3-8c6592087ac9">

```js
Plot.waffleX([10]).plot()
```

If we wanted to allow non-square waffle cells, then we could avoid overlap in this case. But it’s pretty easy to achieve effectively the same result with the **clip** option:

<img width="631" alt="Screenshot 2024-08-04 at 3 02 15 PM" src="https://github.com/user-attachments/assets/dc18f27e-3f56-4cbf-89d2-37cecb587e54">

```js
Plot.waffleX([10]).plot({clip: true})
```

I think we could file a feature request to support non-square waffle cells, and I can take another pass at it now maybe. But it was just a pain because the code is written in such a way to not have duplicate implementations for waffleX and waffleY, so it’s weird keeping track of which dimension is which. And I don’t think it’ll be very common to want a waffle chart when you only have a few points (but it could occur if the data is dynamic).

Reply:
Yes the use case I have in mind is when you want to animate / build up a chart by adding units one by one.

Reply:
Okay, added support for squished waffles.

Reply:
Thanks for the counter example, I hadn't thought of the thin + long case.

A solution for my chart `Plot.waffleY([1, 1, 1, 1]).plot()` is to add an invisible mark such as `Plot.waffleY([4], {fill: []})`.